lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+
+#include "io.h"
+#include "expression.h"
+#include "tools.h"
+#include "../evaluate/fold.h"
+#include "../parser/tools.h"
+
+namespace Fortran::semantics {
+
+// TODO: C1234, C1235 -- defined I/O constraints
+
+using DefaultCharConstantType =
+    evaluate::Constant<evaluate::Type<common::TypeCategory::Character, 1>>;
+
+void IoChecker::Enter(const parser::ConnectSpec &spec) {
+  // ConnectSpec context FileNameExpr
+  if (std::get_if<parser::FileNameExpr>(&spec.u)) {
+    SetSpecifier(SpecifierKind::File);
+  }
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::CharExpr &spec) {
+  SpecifierKind specKind{};
+  using ParseKind = parser::ConnectSpec::CharExpr::Kind;
+  switch (std::get<ParseKind>(spec.t)) {
+  case ParseKind::Access: specKind = SpecifierKind::Access; break;
+  case ParseKind::Action: specKind = SpecifierKind::Action; break;
+  case ParseKind::Asynchronous: specKind = SpecifierKind::Asynchronous; break;
+  case ParseKind::Blank: specKind = SpecifierKind::Blank; break;
+  case ParseKind::Decimal: specKind = SpecifierKind::Decimal; break;
+  case ParseKind::Delim: specKind = SpecifierKind::Delim; break;
+  case ParseKind::Encoding: specKind = SpecifierKind::Encoding; break;
+  case ParseKind::Form: specKind = SpecifierKind::Form; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Position: specKind = SpecifierKind::Position; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  case ParseKind::Convert: specKind = SpecifierKind::Convert; break;
+  case ParseKind::Dispose: specKind = SpecifierKind::Dispose; break;
+  }
+  SetSpecifier(specKind);
+  if (const auto *expr{GetExpr(std::get<parser::ScalarDefaultCharExpr>(spec.t)
+                                   .thing.thing.value())}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      std::string s{parser::ToUpperCaseLetters(**charConst)};
+      if (specKind == SpecifierKind::Access) {
+        hasKnownAccess_ = true;
+        hasAccessDirect_ = s == "DIRECT"s;
+        hasAccessStream_ = s == "STREAM"s;
+      }
+      CheckStringValue(specKind, **charConst, parser::FindSourceLocation(spec));
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::Newunit &) {
+  SetSpecifier(SpecifierKind::Newunit);
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::Recl &spec) {
+  SetSpecifier(SpecifierKind::Recl);
+  if (const auto *expr{GetExpr(spec)}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    const std::optional<std::int64_t> &recl{evaluate::ToInt64(foldExpr)};
+    if (recl && *recl <= 0) {
+      context_.Say(parser::FindSourceLocation(spec),
+          "RECL value (%d) must be positive"_err_en_US,
```

Comment:
The format here should probably be `%jd` and the value `static_cast<std::intmax_t>(*recl)`.

Reply:
Done